synapse/lib/remcycle.py

```diff
@@ -0,0 +1,846 @@
+                       changing the job execution.
+        :return: Job id value which can be referenced against the Hypnos boss
+                 object.
+        :raises: NoSuchName if the API name does not exist.
+        '''
+        # First, make sure the name is good
+        nyx = self.getApiNyx(name=name)
+
+        # Grab things out of kwargs
+        callback = kwargs.pop('callback', None)
+        ondone = kwargs.pop('ondone', None)
+        job_timeout = kwargs.pop('job_timeout', None)
+        fail_fast = kwargs.pop('fail_fast', True)
+        api_args = kwargs.get('api_args', {})
+
+        if not callback:
+            # Setup the default callback
+            def default_callback(*cbargs, **cbkwargs):
+                self.fire(evtname=name, **{'args': cbargs, 'kwargs': cbkwargs})
+
+            callback = default_callback
+        # Wrap the callback so that it will fail fast in the case of a request error.
+        if fail_fast:
+            callback = self._resp_fail_wrapper(callback)
+
+        # Construct the job tufo
+        jid = s_async.jobid()
+        t = s_async.newtask(callback, *args, **kwargs)
+        job = self.boss.initJob(jid, task=t, ondone=ondone, timeout=job_timeout)
+
+        # Create our Async callback function - it enjoys the locals().
+        def response_nommer(resp):
+            job_kwargs = job[1]['task'][2]
+            job_kwargs['jid'] = job[0]
+            if resp.error:
+                _e = resp.error
+                _execinfo = {
+                    'err': _e.__class__.__name__,
+                    'errmsg': str(_e),
+                    'errfile': '',
+                    'errline': '',
+                }
+                job_kwargs['excinfo'] = _execinfo
+            resp_dict = self._flatten_response(resp=resp)
+            job_kwargs['resp'] = resp_dict
+            self.pool.call(self.boss._runJob, job)
+
+        # Construct the request object
+        req = nyx.buildHttpRequest(api_args=api_args)
+
+        self.client.fetch(req, callback=response_nommer)
+
+        return jid
+
+
+def main(argv, outp=None):  # pragma: no cover
+    '''
+    Example usage of remcycle.Hypnos
+    '''
+    if outp is None:  # pragma: no cover
+        outp = s_output.OutPut()
+
+    p = makeargpaser()
+    options = p.parse_args(argv)
+
+    if not options.verbose:
+        logging.disable(logging.DEBUG)
+
+    gconfig = {
+        'apis': {
+            'fqdn': {
+                'doc': 'Get arbitrary domain name.',
+                'http': {
+                    'validate_cert': False
+                },
+                'url': 'https://{{fqdn}}/',
+                'api_args': ['fqdn']
+            }
+        },
+        'doc': 'Definition for getting an arbitrary domain.',
+        'http': {
+            'user_agent': 'SynapseTest.RemCycle'
+        },
+        'namespace': 'generic',
+    }
+
+    fqdns = ['www.google.com',
```

Comment:
this main() function seems like test/example code that should probably live in the test rather than potentially being confused for production code :D